### ISSUE TITLE
How do you access the temporary files before uploading to S3 with skipper?

### ISSUE THREAD

**Issue by albertpeiro** [balderdashy/skipper#54]
Is there any way I can know the names of the the temporary files from req (in an api service) before I upload ?

If there's no way: how do I access req.files. Doesn't seem to work?

Thank you

**Comment by mikermcneil** [comment]
Hey @albertpeiro, thanks for asking this question.  Because it's a complicated answer, I put a lot of thought into it and started putting together a blog post.  I answered your questions inline below for quick reference; but also make sure and also check out the conceptual diagram.  I'll also post a link to the final post when I finish it.

### Is there any way I can know the names of the the temporary files from req (in an api service) before I upload ?

Instead of using the `.upload()` convenience method, you can tap in directly to the [upstream](https://github.com/balderdashy/skipper#what-are-upstreams) of incoming files:

```js
function (req, res) {
  var SkipperDiskAdapter = require('skipper-disk');
  var receiver = SkipperDiskAdapter().receive({/* opts */});
  var upstream = req.file('screenshots`);
  // ... build an intermediate transform stream to do what you need, e.g. `MyIntermediateStream`...
  upstream.pipe(MyIntermediateStream).pipe(receiver);
}
```

In the next release of Skipper, I'm interested in exposing a built-in configurable function you can use to accept or reject any given incoming file from an upstream _before_ it is passed to the receiver (this is effectively how the `maxBytes` option works)  If you're reading this, please let me know if you also have this use case and have time to contribute.  We need your help.

### If there's no way: how do I access req.files. Doesn't seem to work?

So the `req.files` you're used to was a property of the traditional Express/Connect body parser from Express <= v3.0.0.  The way it works is to upload incoming files to a temporary directory on disk -- _before_ running your app-level code.  That means if a user agent is uploading three 100MB files, they must be entirely streamed to the server's hard disk _before_ your code can run.   While it was possible to limit this somewhat in previous versions of Express, it was still a challenge- and more importantly, an easy-to-miss DoS vulnerability.  So Express 4 [removed file upload support](https://github.com/expressjs/body-parser#body-parser) from core.  Meanwhile, @sgress454 and I were working on Skipper because I didn't want to just leave people using Sails (including myself) hanging.

You can use any bodyparser you like w/ Sails or Express (there are lots of good ones!)  In Sails, for example, you can override the default bodyparser middleware (Skipper) by [configuring `sails.config.http.middleware`](http://sailsjs.org/#/documentation/concepts/Middleware?q=adding-or-overriding-http-middleware).  You shouldn't need to do that in 90% of cases, but in case you run into something insurmountable, rest assured you always have the option.


## Body Parsers Through the Ages

### Using legacy Express/Connect body parser
   
```
                                                                          (  use req.files to get the path       )
                                                                         (  to each already-uploaded file on disk )
                                                                         (  on disk, then copy it to a new folder )
                                                                          (   or stream it out to S3, etc.       )
                                                                               |
                                                                               |
                      _________________           __________________           •   _______________________
   file    =~=~=~=~>  |   legacy      | ==/ /==>  |  tmp folder    | ====/ /====>  | your routes/actions |
  uploads             |  body parser  |           | on server disk |               |_____________________|
                      |_______________|           |________________|               |_|_|_|_|_|_|_|_|_|_|_|
```


### With no body parser

```
     ( manually parse req stream  )
      ( in your app-level code   )
     (   on a per-route basis   )
                  |
                  |
                  •   _______________________
   file    =~=~=~=~>  | your routes/actions |
  uploads             |_____________________|
                      |_|_|_|_|_|_|_|_|_|_|_|
```

### Using Skipper

```
                                  ( use req.file('foo') to access upstream )
                                 (  then call .upload() to pipe incoming    )
                                 ( file(s) directly to an upstream receiver )
                                  (  like Amazon S3, Mongo gridfs, or disk )
                                           |
                                           |
                      _________________    •     _______________________
   file    =~=~=~=~>  |   Skipper     | =~=~=~>  | your routes/actions |
  uploads             |  body parser  |          |_____________________|
                      |_______________|          |_|_|_|_|_|_|_|_|_|_|_|
```


### Legend

| Symbol       | Meaning                                                                         |
| ------------ | ------------------------------------------------------------------------------- |
| `=~=~=~=~>`  | incoming file stream, arriving a few kB at a time (don't wait for entire file)  |
| `===/ /==>`  | entire file duplicated on disk (wait for entire file before proceeding)         |

**Comment by mikermcneil** [comment]
@albertpeiro added intercept function discussed above to ROADMAP.md in https://github.com/balderdashy/skipper/commit/9916b4c7db31c7c66caeb7bd6552935c0f4dc3ac

**Comment by sgress454** [comment]
Another oft-overlooked benefit of Skipper is that it enables you to monitor file upload progress on the server (via the `onProgress` event), something that's impossible using schemes where the entire file is uploaded before it becomes available in the request.

**Comment by albertpeiro** [comment]
thanks @mikermcneil for thoroughness of your answer.

My scenario is this:
- API-only in sailsjs with skipper
- Final storage is S3 in a separate vpc.
- Client side app uploads small images (up to 2MB let's say)
- API is capable of validating a set of constraints (width, height, ratio, kilobytes) on the image before it transfers to S3 and return validation errors to client
- (in the future) API is capable of generating thumbnails from the original and upload them in the same spawn of operations.
- End goal is sailsjs API's can be given upload superpowers with relative ease, cheap cost and acceptable performance, including transformations on the files before upload. All with the simplicity flair characteristic of sailsjs.

I'm guessing it's a pretty common use case for a sailsjs or node user. Anyone who's ever done image servers right will appreciate the power of this characteristic in the context of a node /sails env.

**Comment by albertpeiro** [comment]
@sgress454 thanks for sharing that! I wasn't aware. Surely useful at least for logging progress, but I'm sure there'll be other interesting advanced use cases.

**Comment by inklingsofreality** [comment]
Just a thought, but what would the intermediate stream look like if one is using gm().
This is what I have in trying to implement this:
```
myIntermediateStream._transform = function(data, enc, done){
            gm(data).resize('500','','^').gravity('Center').crop('500','500').stream().pipe(data);
            console.log(data);
            done();
        };
```
I really wish there was an easier way to intercept the files before uploading them using an adapter. It seems that this is a necessary feature.

**Comment by sailsbot** [comment]
Thanks for posting, @albertpeiro.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
